webhooks/virtualmachine/mutation/virtualmachine_mutator.go

```diff
@@ -310,24 +311,74 @@ func SetDefaultPowerState(
 	return false
 }
 
-// ResolveImageName mutates the vm.spec.imageName if it's not set to a vmi name
-// and there is a single namespace or cluster scope image with that status name.
+const (
+	vmiKind  = "VirtualMachineImage"
+	cvmiKind = "Cluster" + vmiKind
+)
+
+// ResolveImageName ensures vm.spec.image is set to a non-empty value if
+// vm.spec.imageName is also non-empty.
 func ResolveImageName(
 	ctx *context.WebhookRequestContext,
 	c client.Client,
 	vm *vmopv1.VirtualMachine) (bool, error) {
-	// Return early if the VM image name is empty or already set to a vmi name.
-	imageName := vm.Spec.ImageName
-	if imageName == "" || strings.HasPrefix(imageName, "vmi-") {
+
+	// Return early if the image reference is used.
+	if vm.Spec.Image != nil {
+		return false, nil
+	}
+
+	// Return early if the VM image name is empty.
+	imgName := vm.Spec.ImageName
+	if imgName == "" {
 		return false, nil
 	}
 
-	var determinedImageName string
+	// Query the image from the object name in order to set the result in
+	// spec.image.
+	if strings.HasPrefix(imgName, "vmi-") {
+		vm.Spec.Image = &vmopv1.VirtualMachineImageRef{
+			Name: imgName,
+		}
+
+		if err := c.Get(
+			ctx,
+			client.ObjectKey{Namespace: vm.Namespace, Name: imgName},
+			&vmopv1.VirtualMachineImage{}); err != nil {
+
+			if !apierrors.IsNotFound(err) {
+				return false, err
+			}
+
+			if err := c.Get(
+				ctx,
+				client.ObjectKey{Name: imgName},
+				&vmopv1.ClusterVirtualMachineImage{}); err != nil {
+
+				return false, err
```

Comment:
We might want to add some additional error messages here for better clarity. Otherwise, if a user specifies a non-existing "vmi-" image name, expecting it to be a namespace scope image, they would always see "failed to get ClusterVirtualMachineImage" returned from the current `err`.

Reply:
Fair. It used to be more general as well. Let me address that.

Reply:
I was thinking of a different location that used to have a more generic error. In this instance, I think this is probably okay:

> Otherwise, if a user specifies a non-existing "vmi-" image name, expecting it to be a namespace scope image

If it is non-existing, then they won't know what type to expect. If they expected it to be namespace scoped, they could have simply provided the actual kind in `spec.image`. 

I may add a message that indicates neither a namespace nor cluster scoped resource could be found, but still, this is probably fine as-is. Still, it's a good catch.

Reply:
@dilyar85 I updated the error, PTAL.

Reply:
Thanks for addressing this.